checker/src/call_graph.rs

```diff
@@ -0,0 +1,770 @@
+use petgraph::dot::{Config, Dot};
+use petgraph::graph::{DefaultIx, NodeIndex};
+use petgraph::visit::Bfs;
+use petgraph::{Direction, Graph};
+use rustc_hir::def_id::DefId;
+use serde::{Deserialize, Serialize};
+use serde_json;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::fs;
+use std::path::Path;
+
+// An unique identifier for a Rust type string.
+type RTypeIdx = u32;
+
+// A unique identifier for a graph node.
+type NodeIdx = NodeIndex<DefaultIx>;
+
+// Convenience types for the graph folding algorithm.
+type HalfRawEdge = (NodeIdx, RTypeIdx);
+type RawEdge = (NodeIdx, NodeIdx, RTypeIdx);
+type MidpointExcludedMap = HashMap<NodeIdx, (HashSet<HalfRawEdge>, HashSet<HalfRawEdge>)>;
+
+/// Specifies reduction operations that may be perfomed
+/// on a call graph. Supported operations are:
+#[derive(Debug, Clone, Serialize, Deserialize)]
+enum CallGraphReduction {
+    /// Only include nodes reachable from the given node.
+    /// See `CallGraph::filter_reachable`.
+    Slice(String),
```

Comment:
this should probably be str. Have you run Clippy? I.e. run validate.sh before putting up a PR>

Reply:
I'll change that to be a str.

I did run Clippy (`cargo clippy --workspace`) and did not receive any warnings. I didn't run validate.sh at the time, but am running it now.